In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
from taxonomy_generators import build_tools, build_registry


tools = build_tools()
registry = build_registry()

In [3]:
from query_taxonomy.features import FeatureGroup


registry[FeatureGroup.STRUCTURED_IDENTIFIERS][0]

In [11]:
from litellm import completion
import instructor

response = completion(
  model="anthropic/claude-haiku-4-5-20251001",
  messages=[{"role": "user", "content": "Hello, how are you?"}]
)
print(response.choices[0].message.content)

Hello! I'm doing well, thanks for asking. I'm here and ready to help with whatever you need. How are you doing today?


In [8]:
import json 

from taxonomy_generators import build_tools, catalog

TOOLS = build_tools(seed=0)
RUNNERS = { spec.name: spec.run for spec in TOOLS }

CLAUDE_TOOLS = [
    {"name": s.name, "description": s.description, "input_schema": s.input_schema}
    for s in TOOLS
]

SYSTEM = """You enrich search queries with taxonomy features.
Given a sentence and target feature names:
1. call generate_surface for each requested feature,
2. weave the surfaces into the sentence so it stays a plausible search query,
3. call verify on your candidate with span targets (min_count=1 per feature),
4. if verify fails, adjust the text and verify again.
End with ONLY the final enriched sentence, nothing else."""


In [18]:
import json

import instructor
from litellm import completion
from pydantic import BaseModel, Field


class EnrichedQuery(BaseModel):
    text: str = Field(description="The enriched sentence that passes verify")
    features_used: list[str] = Field(description="Feature names woven in")


class Enricher:
    def __init__(self, model="anthropic/claude-haiku-4-5-20251001"):
        self.model = model
        self.client = instructor.from_litellm(completion)

    def enrich(self, sentence: str, features: list[str]) -> EnrichedQuery:
        messages = [
            {
                "role": "system",
                "content": SYSTEM,
            },
            {
                "role": "user",
                "content": (
                    f"Sentence: {sentence}\n"
                    f"Features: {', '.join(features)}"
                ),
            },
        ]

        while True:
            response = completion(
                model=self.model,
                messages=messages,
                tools=CLAUDE_TOOLS,
                tool_choice="auto",
                max_tokens=1024,
            )

            message = response.choices[0].message
            messages.append(message.model_dump())

            tool_calls = getattr(message, "tool_calls", None)
            if not tool_calls:
                break

            for call in tool_calls:
                args = json.loads(call.function.arguments)

                result = RUNNERS[call.function.name](**args)

                messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": call.id,
                        "content": json.dumps(result),
                    }
                )

        return self.client.chat.completions.create(
            model=self.model,
            response_model=EnrichedQuery,
            messages=messages
            + [
                {
                    "role": "user",
                    "content": "Return the final verified enrichment as structured output.",
                }
            ],
        )

In [19]:
print([entry["feature"] for entry in catalog()][:20])

['cve', 'datetime', 'cidr', 'ip_address', 'mac_address', 'uri', 'email', 'api_key', 'uuid', 'env_var', 'hex_color', 'standards_citation', 'currency_amount', 'iban', 'lei', 'industry_code', 'business_registration', 'legal_citation', 'court_docket', 'neutral_citation']


In [22]:
enricher = Enricher()
enricher.enrich('How to a get to this website', ['ip_address', 'api_key'])

EnrichedQuery(text='How to get to this website at 0.4.251.255 using API key ghp_rb3wt2C6IpcgmVKuRVWAaIHdZN3OHaSalxg7CWsGb0uHqVATwWjFmoRGg9zI0tL5', features_used=['ip_address', 'api_key'])

In [23]:
enricher = Enricher()
enricher.enrich('Conforting home for 200 euro/month', ['datetime'])

EnrichedQuery(text='Conforting home for 200 euro/month since 1523759459', features_used=['datetime'])

In [ ]:
enricher = Enricher()
enricher.enrich('Blue is the color of month', ['hex_color'])

EnrichedQuery(text='Blue #cdE20D is the color of month', features_used=['hex_color'])

In [ ]:
from taxonomy_generators import SpanTarget, Targets, verify


features = ["ip_address", "api_key"]
result = enricher.enrich("How to a get to this website", features)

In [28]:
report = verify(result.text, Targets(spans=tuple(SpanTarget(feature=f) for f in features)))

In [30]:
print("passed: ", report.passed)

passed:  True


In [31]:
for check in report.checks:
    print(f"  {check.target}: measured={check.measured}")

  ip_address: at least 1 span(s): measured=1.0
  api_key: at least 1 span(s): measured=1.0


## Augmentation part 

Lets check how the pseudo-code was implemented:
```python
while hungry_features_exist(order_sheet):
    feature = select_most_hungry(order_sheet)            # a span floor OR a stat band

    if is_stat(feature):                                 # e.g. length_words:60+
        band  = hungry_band(feature)                     # the thin part of the distribution
        query = find_query_movable_into(band)            # from the STATS table: short, zero-span
        new_query, ok = rewrite_towards(query, band)     # LLM expands / compresses / restructures
        answer = answer_of(query)                        # meaning kept -> answer unchanged

    elif needs_document(feature):                        # identifiers
        query, doc, surface = find_pair(feature)         # SPANS table: doc contains the surface
                                                         # + STATS table: query profile fits
        new_query, ok = enrich(query, surface)
        answer = doc

    else:                                                # markers, operator_syntax
        query = find_query(feature)                      # eligibility is often a STAT rule
        new_query, ok = enrich(query, feature)
        answer = answer_of(query)

    ok = ok and verify(new_query, feature)               # spans: feature present
                                                         # stats: scalar landed in the band
                                                         # both: no unwanted new spans
    if ok: write(new_query, answer)
```

### Setup the environment

In [4]:
import pandas as pd 
from augmentation import AugmentationLoop, Augmenter, GeneratedPool, operator_for
from composition import TargetComposition

selection = TargetComposition().build()

loop = AugmentationLoop(selection)

/Users/andrei/projects/hybrid-search-rrf-dataset/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [5]:
# hungry_features
loop.hungry()

,floor,missing,operator,gate,runnable
0,id:datetime,997.00,NaN,NaN,False
1,id:medical,992.50,NaN,NaN,False
2,logical:operator_syntax,965.00,NaN,NaN,False
3,id:logistics,890.50,NaN,NaN,False
4,id:media,875.75,NaN,NaN,False
5,marker:greeting,861.00,decorate,none,True
6,marker:interjection,853.00,decorate,none,True
7,id:network,433.75,NaN,NaN,False
8,length_words:60+,426.00,NaN,NaN,False
9,id:shape_guess,135.75,NaN,NaN,False


In [6]:
loop.order_sheet()

,slice,floor,amount,credit,missing,reason
0,A,id:datetime,1000.0,3.00,997.00,exhausted
1,A,id:medical,1000.0,7.50,992.50,exhausted
2,A,logical:operator_syntax,1000.0,35.00,965.00,exhausted
3,A,id:logistics,1000.0,109.50,890.50,exhausted
4,A,id:media,1000.0,124.25,875.75,exhausted
5,A,marker:greeting,1000.0,139.00,861.00,exhausted
6,A,marker:interjection,1000.0,147.00,853.00,exhausted
7,A,id:network,1000.0,566.25,433.75,capped
8,C,length_words:60+,714.0,288.00,426.00,exhausted
9,A,id:shape_guess,1000.0,864.25,135.75,exhausted


In [7]:
declaration = operator_for('marker:politeness').declaration

print(declaration)

operator='decorate' floor_prefix='marker:' grounding=<Grounding.NONE: 'none'> answer_key=<AnswerKeyPath.INHERIT: 'inherit'> meaning_preserved=True verifiable_by='target marker span present on local re-measure' credit_gate=<CreditGate.NONE: 'none'>


In [8]:
operator = operator_for("marker:acronym")

operator

In [9]:
operator = operator_for("id:datetime")

operator

In [10]:
row = selection.iloc[0]
row.query, list(row.floors)

('DHA', ['id:shape_guess', 'marker:acronym'])

In [11]:
op = operator_for("marker:politeness")
parents = op.eligible(selection, "marker:politeness")

len(parents)

32871

In [12]:
op.instruction("marker:politeness")

"Rewrite the user's search query by weaving in a politeness phrase (e.g. 'please', 'could you kindly'). Write the phrase yourself — no generator call is needed. Keep every content word and the meaning unchanged. Add no other information: no names, numbers, dates, or identifiers. Check your text with the verify tool (span feature 'politeness'), then return the final query text."

In [13]:
op.targets("marker:politeness")

Targets(spans=(SpanTarget(feature='politeness', min_count=1, max_count=None),), stats=())

In [14]:
engine = Augmenter()
out = engine.run(op.instruction("marker:politeness"), "Query: what does DHA do", op.targets("marker:politeness"))

out.text, out.accepted, out.attempts

('Could you please explain what DHA does?', True, 1)

In [18]:
engine = Augmenter()
out = engine.run(op.instruction("marker:interjection"), "Query: I went to speak with my colleague and found out...", op.targets("marker:negation"))

out.text, out.accepted, out.attempts

('Oh, I went to speak with my colleague and found out he could not help...',
 True,
 1)

| feature (floor)     | hungry credits | status       |
| ------------------- | -------------: | ------------ |
| marker:politeness   |             14 | runnable now |
| marker:greeting     |            861 | runnable now |
| marker:interjection |            853 | runnable now |

In [20]:
t = op.targets("marker:politeness")
engine.accept("could you please explain what DHA does", t).passed  == True # True
engine.accept("what does DHA do", t)                                # passed=False, checks name the miss

VerifyReport(passed=False, checks=(TargetCheck(target='politeness: at least 1 span(s)', measured=0.0, passed=False),))

In [21]:
from taxonomy_generators.verify import SpanTarget, Targets

two_polite = Targets(spans=(SpanTarget(feature="politeness", min_count=2),))

out = engine.run(
    "Rewrite the user's search query by weaving in TWO distinct politeness "
    "phrases (e.g. 'please' and 'could you kindly'). Keep every content word "
    "and the meaning unchanged; add no other information. Check with the "
    "verify tool, then return the final query text.",
    "Query: what does DHA do",
    two_polite,
)

out

AugmentationOutcome(text='Please tell me what does DHA do, could you kindly?', accepted=True, attempts=1, checks=(TargetCheck(target='politeness: at least 2 span(s)', measured=3.0, passed=True),))

In [22]:
from composition import TargetComposition
from augmentation import AugmentationLoop

loop = AugmentationLoop(TargetComposition().build())
loop.hungry()                  

,floor,missing,operator,gate,runnable
0,id:datetime,997.00,NaN,NaN,False
1,id:medical,992.50,NaN,NaN,False
2,logical:operator_syntax,965.00,NaN,NaN,False
3,id:logistics,890.50,NaN,NaN,False
4,id:media,875.75,NaN,NaN,False
5,marker:greeting,861.00,decorate,none,True
6,marker:interjection,853.00,decorate,none,True
7,id:network,433.75,NaN,NaN,False
8,length_words:60+,426.00,NaN,NaN,False
9,id:shape_guess,135.75,NaN,NaN,False


In [ ]:
loop.run("marker:politeness")

augment:marker:politeness:   7%|▋         | 1/14 [00:07<01:42,  7.85s/row, attempted=1, dropped=0]

+ 4083 (1 attempt(s)): 'Could you please help me solve this problem: The only difference between easy an'


augment:marker:politeness:  14%|█▍        | 2/14 [00:12<01:11,  6.00s/row, attempted=2, dropped=0]

+ 1098731 (1 attempt(s)): 'Could you please tell me how long leftovers are good for in the refrigerator?'


augment:marker:politeness:  21%|██▏       | 3/14 [00:16<00:56,  5.12s/row, attempted=3, dropped=0]

+ 902420 (1 attempt(s)): 'Could you tell me what temperature to keep your house at for ideal sleep?'


augment:marker:politeness:  29%|██▊       | 4/14 [00:20<00:46,  4.65s/row, attempted=4, dropped=0]

+ test-992 (1 attempt(s)): 'Could you kindly show me the flora of Suriname not Brazil'


augment:marker:politeness:  36%|███▌      | 5/14 [00:25<00:41,  4.60s/row, attempted=5, dropped=0]

+ 4716 (1 attempt(s)): 'Could you please tell me if mail is a primary method of service for eviction act'


KeyboardInterrupt: 

In [25]:
loop.run("marker:greeting")

                                                                                                  
augment:marker:politeness:  36%|███▌      | 5/14 [02:03<00:41,  4.60s/row, attempted=5, dropped=0]

+ 984840 (1 attempt(s))
    before: 'why can pregnant women be mean'
    after:  'Hi there, why can pregnant women be mean'


                                                                                                  
augment:marker:politeness:  36%|███▌      | 5/14 [02:07<00:41,  4.60s/row, attempted=5, dropped=0]

+ MATH-q-4344 (1 attempt(s))
    before: 'Problem: Consider the rectangle with vertices at $(5,4),$ $(5,-4),$ $(-5,4),$ $('
    after:  'Hi there, consider the rectangle with vertices at $(5,4),$ $(5,-4),$ $(-5,4),$ $'


                                                                                                  
augment:marker:politeness:  36%|███▌      | 5/14 [02:12<00:41,  4.60s/row, attempted=5, dropped=0] 

+ MATH-q-2008 (1 attempt(s))
    before: 'Problem: If $j$, $k$, and $l$ are positive with $jk=24$, $jl = 48$, and $kl=18$,'
    after:  'Hi there, Problem: If $j$, $k$, and $l$ are positive with $jk=24$, $jl = 48$, an'


                                                                                                  
augment:marker:politeness:  36%|███▌      | 5/14 [02:17<00:41,  4.60s/row, attempted=5, dropped=0] 

+ 1078330 (1 attempt(s))
    before: 'what important distinction did president john f. kennedy make regarding civil ri'
    after:  'Hi there, what important distinction did president john f. kennedy make regardin'


                                                                                                  
augment:marker:politeness:  36%|███▌      | 5/14 [02:21<00:41,  4.60s/row, attempted=5, dropped=0] 

+ 457076 (1 attempt(s))
    before: 'most common technology used in classroom'
    after:  'Hi there, most common technology used in classroom'


KeyboardInterrupt: 